In [45]:
from html.parser import HTMLParser as hp
import html
import re
import copy
import numpy as np
import pandas as pd
import json
import glob
from random import shuffle
from bs4 import BeautifulSoup as bs

### file read

In [25]:
data_path = "data"
#data_path = "data(test)"

file_list = glob.glob("%s/*.json" % data_path)

json_train=[]

shuffle(file_list)
for json_file_name in file_list:
    print(json_file_name)
    json_file = json.loads(open(json_file_name).read())
    json_train += json_file["articles"]
    
print(len(json_train))

data/14.json
data/7.json
data/8.json
data/10.json
data/4.json
data/0.json
data/11.json
data/1.json
data/13.json
data/16.json
data/9.json
data/17.json
data/12.json
data/6.json
data/5.json
data/3.json
data/15.json
data/2.json
172044


### get tag type & make DF

In [61]:
tag_list = []
label_list = []

num_json_train = len(json_train)

for i, article in enumerate(json_train):
    if i % 10000 == 0:
        print("converting... (%d/%d)" % (i, num_json_train))

    tags = {}

#     train_text = article["text"]
#     matches = re.findall('<[^/].*?>', train_text)
    
#     for tag in matches:
#         tag_name = 'TAG_' + tag.replace('<','').replace('>','').strip().split(' ')[0]
#         #print(tag_name)
#         if tag_name not in tags:
#             tags[tag_name] = 0
#         tags[tag_name] += 1
    
    train_text = article["text"]
    soup = bs(train_text, 'html.parser')
    for tag in soup.find_all():
        tag_name = 'TAG_' + tag.name
        if tag_name not in tags:
            tags[tag_name] = 0
        tags[tag_name] += 1

    tags["is_troll"] = article["is_troll"]
    tag_list.append(tags)
    

data = pd.DataFrame.from_dict(tag_list)
data = data.fillna(0)

data

converting... (0/172044)
converting... (10000/172044)
converting... (20000/172044)
converting... (30000/172044)
converting... (40000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://ilwar.com/free/44658'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://ilwar.com/free/45344'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautif


converting... (50000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:198: UserWarning: "b'..'" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.
  '"%s" looks like a filename, not markup. You should probably open this file and pass the filehandle into Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://blog.naver.com/kakashi08/110086751501'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserW


converting... (60000/172044)
converting... (70000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://image.kilho.net/?pk=1443527\r\n\r\n\r\n\r\nFree'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://www.youtube.com/watch?v=ByhP478ZPQI\r\n\r\n\xe3\x84\xb9\xe3\x85\x87\xec\xa1\xb4\xec\xa2\x8b'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should


converting... (80000/172044)
converting... (90000/172044)
converting... (100000/172044)
converting... (110000/172044)
converting... (120000/172044)
converting... (130000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://imgur.com/rBtm1LI.jpg'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.ilwar.com/poli/17062'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beau


converting... (140000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://www.ilbe.com/2691808050'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://withgonggam.tistory.com/m/post/view/id/1316'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and fee


converting... (150000/172044)
converting... (160000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.khan.co.kr/view.html?category=2&med_id=khan&artid=201401071905352&code=361102'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://news.sportsseoul.com/read/entertain/1295201.htm<div><span><br><br>\xed\x95\x98\xec\x95\x84...</span></div>'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. B


converting... (170000/172044)

/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.todayhumor.co.kr/view.php?table=humordata&no=1488717\r<br>\r<br>http://m.blog.naver.com/zerotimer/100202776978'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.' % markup)
/Library/Frameworks/Python.framework/Versions/3.4/lib/python3.4/site-packages/bs4/__init__.py:207: UserWarning: "b'http://m.media.daum.net/m/media/politics/newsview/20140105055803492'" looks like a URL. Beautiful Soup is not an HTTP client. You should probably use an HTTP client to get the document behind the URL, and feed that document to Beautiful Soup.
  '"%s" looks like a URL. Beautiful Soup

TAG_a  TAG_ahahahahh  TAG_angangangangang  TAG_audio  TAG_b  \
0           0              0                    0          0      0   
1           0              0                    0          0      0   
2           0              0                    0          0      0   
3           7              0                    0          0     39   
4           1              0                    0          0      0   
5           0              0                    0          0      0   
6           2              0                    0          0      0   
7           0              0                    0          0      0   
8           0              0                    0          0      1   
9           0              0                    0          0      0   
10          0              0                    0          0      0   
11          1              0                    0          0      0   
12          0              0                    0          0      0   
13          0              0                    0          0      0   
14          3              0                    0          0     53   
15          1              0                    0          0      0   
16          0              0                    0          0      0   
17          0              0                    0          0      0   
18          0              0                    0          0      0   
19          0              0                    0          0      0   
20          1              0                    0          0      0   
21          0              0                    0          0      0   
22          0              0                    0          0      0   
23          0              0                    0          0      0   
24          0              0                    0          0      0   
25          0              0                    0          0      0   
26          0              0                    0          0      1   
27          0              0                    0          0      0   
28          0              0                    0          0      0   
29          0              0                    0          0      0   
...       ...            ...                  ...        ...    ...   
172014      0              0                    0          0      0   
172015      0              0                    0          0      0   
172016      0              0                    0          0      0   
172017      0              0                    0          0      0   
172018      0              0                    0          0      0   
172019      0              0                    0          0      0   
172020      0              0                    0          0      0   
172021      0              0                    0          0      0   
172022      0              0                    0          0      0   
172023      0              0                    0          0      0   
172024      0              0                    0          0      0   
172025      0              0                    0          0      0   
172026      0              0                    0          0      0   
172027      0              0                    0          0      0   
172028      0              0                    0          0      0   
172029      0              0                    0          0      0   
172030      1              0                    0          0      0   
172031      0              0                    0          0      0   
172032      0              0                    0          0      0   
172033      0              0                    0          0      0   
172034      0              0                    0          0      0   
172035      0              0                    0          0      0   
172036      0              0                    0          0      0   
172037      0              0                    0          0      0   
172038      0              0    

In [34]:
data

TAG_  TAG_!--  TAG_"http://i.imgur.com/3glK3Lh.png"  TAG_'  TAG_-  \
0          0        0                                     0      0      0   
1          0        0                                     0      0      0   
2          0        0                                     0      0      0   
3          0        0                                     0      0      0   
4          0        0                                     0      0      0   
5          0        0                                     0      0      0   
6          0        0                                     0      0      0   
7          0        0                                     0      0      0   
8          0        0                                     0      0      0   
9          0        0                                     0      0      0   
10         0        0                                     0      0      0   
11         0        0                                     0      0      0   
12         0        0                                     0      0      0   
13         0        0                                     0      0      0   
14         0        0                                     0      0      0   
15         0        0                                     0      0      0   
16         0        0                                     0      0      0   
17         0        0                                     0      0      0   
18         0        0                                     0      0      0   
19         0        0                                     0      0      0   
20         0        0                                     0      0      0   
21         0        0                                     0      0      0   
22         0        0                                     0      0      0   
23         0        0                                     0      0      0   
24         0        0                                     0      0      0   
25         0        0                                     0      0      0   
26         0        0                                     0      0      0   
27         0        0                                     0      0      0   
28         0        0                                     0      0      0   
29         0        0                                     0      0      0   
...      ...      ...                                   ...    ...    ...   
172014     0        0                                     0      0      0   
172015     0        0                                     0      0      0   
172016     0        0                                     0      0      0   
172017     0        0                                     0      0      0   
172018     0        0                                     0      0      0   
172019     0        0                                     0      0      0   
172020     0        0                                     0      0      0   
172021     0        0                                     0      0      0   
172022     0        0                                     0      0      0   
172023     0        0                                     0      0      0   
172024     0        0                                     0      0      0   
172025     0        0                                     0      0      0   
172026     0        0                                     0      0      0   
172027     0        0                                     0      0      0   
172028     0        0                                     0      0      0   
172029     0        0                                     0      0      0   
172030     0        0                                     0      0      0   
172031     0        0                                     0      0      0   
172032     0        0                                     0      0      0   
172033     0        0                                     0      0      0   
1720

### 1번만 사용되는 tag 제거

In [62]:
drop_col_list = []
ex = []
for colname in data.columns.drop([label_name]):
#    print(colname, sum(data[colname]))
    if sum(data[colname]) == 1:
        drop_col_list.append(colname)
    else:
        ex.append(colname)
        
len(drop_col_list)
ex

['TAG_a',
 'TAG_b',
 'TAG_blockquote',
 'TAG_body',
 'TAG_br',
 'TAG_caption',
 'TAG_center',
 'TAG_cite',
 'TAG_code',
 'TAG_col',
 'TAG_colgroup',
 'TAG_dd',
 'TAG_del',
 'TAG_div',
 'TAG_dl',
 'TAG_dt',
 'TAG_em',
 'TAG_embed',
 'TAG_font',
 'TAG_h1',
 'TAG_h2',
 'TAG_h3',
 'TAG_h4',
 'TAG_h5',
 'TAG_h6',
 'TAG_hr',
 'TAG_i',
 'TAG_iframe',
 'TAG_image',
 'TAG_img',
 'TAG_img<',
 'TAG_jinlove',
 'TAG_jtbc',
 'TAG_kbs',
 'TAG_kkkkkkkk',
 'TAG_li',
 'TAG_marquee',
 'TAG_mbc',
 'TAG_meta',
 'TAG_mianhada',
 'TAG_nicejintime',
 'TAG_object',
 'TAG_ol',
 'TAG_p',
 'TAG_param',
 'TAG_pre',
 'TAG_q',
 'TAG_real',
 'TAG_s',
 'TAG_sbs',
 'TAG_span',
 'TAG_src',
 'TAG_strike',
 'TAG_strong',
 'TAG_sup',
 'TAG_table',
 'TAG_tbody',
 'TAG_td',
 'TAG_tfoot',
 'TAG_th',
 'TAG_thead',
 'TAG_tr',
 'TAG_tv',
 'TAG_u',
 'TAG_ul',
 'TAG_x',
 'TAG_x-script']

In [30]:
data = data.drop(drop_col_list, 1)

### Random Forest

In [63]:
from sklearn.ensemble import RandomForestClassifier as RF

label_name = "is_troll"
predictors = data.columns.drop([label_name])

model = RF(n_estimators=10, n_jobs=4)

### Cross Validation

In [64]:
from sklearn import cross_validation

cv_value = 3

scores = cross_validation.cross_val_score(model, data[predictors], data[label_name], cv=cv_value, scoring='roc_auc')
cv_result = scores.mean()

print(cv_result)

0.644000362278


### Results
beautifulsoup에서 잡아주는 tag만 : 0.644000362278

이상한 형태의 tag 함께 : 0.641983931703

1번 나온 tag (이상한 태그중에서도 한번만 쓰인 것) : 0.645103902784
